dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h

```diff
@@ -223,24 +223,36 @@ class SliceFlipNormalizePermutePadGpu {
       // We fuse the last dimension with the previous IF:
       // 1. There are at least 2 dimensions
       // 2. Last dimension is not sliced/padded/permuted
-      // 3. Last dimension is not the channel dimension
-      // 4. The last two dimensions are either both flipped or not flipped
+      //    - if out_stride[last_dim] == abs(in_stride[last_dim]), then it is not permuted
+      //    - if anchor[last_dim] == 0 && out_shape[last_dim] == in_shape[last_dim],
+      //      then it is not sliced/padded
+      // 3. Neither of the two dimensions to be merged are the channel dimension
+      // 4. The in_strides of the last dimension and the former one align
+      //    - in_strides[last_dim] * in_shape[last_dim] == in_strides[last_dim - 1]
+      //      (this check also makes sure that we collapse dims that have the same sign of strides
+      //       meaning we can merge two dimensions if they are either not flipped or both of them
+      //       are)
+
       int last_dim = Dims - 1;
       while (last_dim > 0 &&
              sample_desc.out_strides[last_dim] ==
-                 static_cast<uint64_t>(sample_desc.in_strides[last_dim]) &&
+                 static_cast<uint64_t>(abs(sample_desc.in_strides[last_dim])) &&
              sample_desc.anchor[last_dim] == 0 &&
              sample_desc.out_shape[last_dim] == sample_desc.in_shape[last_dim] &&
              sample_desc.channel_dim != last_dim &&
-             (sample_desc.in_strides[last_dim] < 0) == (sample_desc.in_strides[last_dim - 1] < 0)) {
+             sample_desc.channel_dim != (last_dim-1) &&
+             sample_desc.in_strides[last_dim] * sample_desc.in_shape[last_dim] ==
+                 sample_desc.in_strides[last_dim - 1]) {
```

Comment:
This condition is scary, was scary before; Could you please outline it to separate function? As it is, it is incomprehensible for me.